### PR TITLE
Misc. bugfixes and enhancements from Alpha Griffin.

### PR DESCRIPTION
**Pull request by lannocc**

### CODE REVIEW

**Comment by teran-mckinney** [comment]
Wow, this is great. Thank you!

Do tests pass? Could probably push this out fairly soon. Would want to validate everything on my end is working as expected.

@excalibur0922, would you mind also taking a look at this?

**Comment by lannocc** [comment]
I have not tried running the test cases. I'll take a look at that. If all tests previously passed then I am guessing a new test will have to be written since this function was previously essentially disabled.

**Comment by lannocc** [comment]
I've now run the test cases. I don't get 100% coverage on the @sporestack HEAD either so no, all tests do not pass. Test results are not consistent as I'm getting 30-32 misses during each test run on @AlphaGriffin and 33-35 misses on @sporestack, possibly due to fluctuating network availability.

**Comment by excalibur0922** [comment]
I didn't have a chance to really look at what you did for OP_RETURN. I think I mis-interpreted (glancing at it on my phone at work)... I think what you've done is functionally, exactly the same as what I've done - if so, yours is more elegant. That's all I wanted. The ability to shove random bytes into that "message" hex / utf-8 or otherwise encoding...

**Comment by lannocc** [comment]
Thanks for the feedback @excalibur0922! Yeah, this really is just a simple enhancement. The gist of it is I add a `bytes` boolean flag to the main transaction-building functions that, when set to true, skips the UTF-8 encoding step and simply assumes you are already providing a byte-string. This flag defaults to False for backwards compatibility. That takes care of the insertion.

To ease the extraction of OP_RETURN (and some other transaction data) I created two new classes called Transaction and TxPart, and created a new function get_transaction() in the network services file that converts the transaction JSON into these proper Python classes. It returns a single instance of Transaction which itself contains instances of TxPart for each of the inputs and outputs. The OP_RETURN data may be retrieved as-is (a hex string as it was returned in the JSON) or there is a simple message() function that will attempt to decode it as UTF-8 if that is desired. But there are no other attempts to interpret the meaning of the data as I believe that is outside the scope of this project.

**Comment by lannocc** [comment]
I also extended the `MESSAGE_SIZE` to 220 bytes, since that is what I've read is the current limit. However, I have not yet sent a transaction with data large enough to reach the limit so this is still untested.

**Comment by excalibur0922** [comment]
Right so I have had a play around. Actually, what I have done IS a bit different.

I want the ability to make: 
6a 02 6d01 09 626974505553484552 (i.e. break up the pushdata into separate chunks... 2 bytes then 9 bytes. 

Using your system I get:
6a 0d 026d0109626974505553484552 (i.e. 0x0d bytes as one big chunk)

doesn't work for memo.cash and is not as versatile. That's cool though :) 

Where I am focusing is on the "construct_output_block" part. giving full control of what goes after OP_RETURN byte by byte

I can just see if Teran is willing to add it separately or after this one. :) I might even see if I can bang it out quick now (because what I'm after is just a single one-and-done patch).

**Comment by lannocc** [comment]
Ah, cool. If I understand you correctly, your system makes it easy to prepend variable-sized chunks with the size of the chunk that follows, and provides a way to read it back out in those same chunks too? That's pretty cool. I could definitely see where that would be useful. It seems generic enough too that I wouldn't find it out of place in a project like this, especially if it's a common scheme already used by many other projects.

For comparison, I happen to be using a separator byte for my variable-sized data only because the variable portions happens to be text and I can use a separator (in my case, 0x00) that I ensure will never exist in the text. But if I needed variable binary data then what you describe is exactly what I would want.

What would be even more powerful is an encoding that compacts at the bit level, instead of whole bytes, to get more use out of the space. For example if I needed an integer of only 3 bits following by an integer of 7 bits, etc. But I can't be bothered to go that far just yet :-).

**Comment by excalibur0922** [comment]
I haven't done it for multiple OP_RETURNS in one transaction at this stage because I think that is non-standard anyway at the moment? But yeah that would be the ultimate... Just ramming a really long string of bytes >> 220 bytes long and having it all spat out into multiple OP_RETURNS within the same txn.
- I have only tackled the case of fine control over a single OP_RETURN of up to 220 bytes max for now...

**Comment by lannocc** [comment]
It's my understanding that BitCash already breaks up anything over MESSAGE_SIZE into multiple outputs, though it does not provide a mechanism to automatically join them back together on read.

**Comment by excalibur0922** [comment]
Yeah multiple OP_RETURN messages within the same transaction... I've never tried relaying anything like that but I thought it was considered "non-standard" by miners? 
Anyway... I haven't dealt with how to fine tune byte by byte when it overflows >220 bytes... That's too much effort for me! haha

**Comment by lannocc** [comment]
Could be. I'll have to take a closer look at what you've done. I'm really just brainstorming though, I don't have specific use cases that need this yet. I'm thinking it's kind of 6 in one hand, half-dozen in the other on whether the caller breaks up large messages and/or does special encoding ahead of time vs passing it in as one large blob along with a callback routine to handle to handle it.

**Comment by teran-mckinney** [comment]
Oh wow, you two have really being going at it. This is great, I'm just not sure what to make of it.

Regarding #10, I can definitely push through a the custom_pushdata feature here. I also think I can help review your commits to bitpusher and having it separate makes the most sense. So hopefully, that ends up being the best of both worlds. It keeps bitcash simpler and more focused.

Should I merge this or #12 first? Also, if you want me to merge #12, please squash it. 50 commits for one PR is too much :).

File: bitcash/network/services.py [hunk 1]
```diff
@@ -32,20 +35,46 @@ def get_balance(cls, address):
...
-    def get_tx_amount(cls, txid, txindex):
-        endpoint = cls.MAIN_TX_AMOUNT_API
+    def get_transaction(cls, txid):
+        r = requests.get(cls.MAIN_TX_API.format(txid), timeout=DEFAULT_TIMEOUT)
+        if r.status_code != 200:  # pragma: no cover
+            raise ConnectionError
+        response = r.json(parse_float=Decimal)
+
+        tx = Transaction(response['txid'], response['blockheight'],
+                (Decimal(response['valueIn'])*100000000).normalize(),
```
Comment (teran-mckinney) [inline bitcash/network/services.py]: Let's use a satoshi constant in these. Would be so easy to have a missing zero and not know.

File: bitcash/network/services.py [hunk 2]
```diff
@@ -73,11 +102,12 @@ class CashExplorerBitcoinDotComAPI(InsightAPI):
     No testnet, sadly. Also uses legacy addresses only.
     """
     MAIN_ENDPOINT = 'https://cashexplorer.bitcoin.com/api/'
-    MAIN_ADDRESS_API = MAIN_ENDPOINT + 'addr/'
-    MAIN_BALANCE_API = MAIN_ADDRESS_API + '{}/balance'
-    MAIN_UNSPENT_API = MAIN_ADDRESS_API + '{}/utxo'
+    MAIN_ADDRESS_API = MAIN_ENDPOINT + 'addr/{}'
```
Comment (teran-mckinney) [inline bitcash/network/services.py]: Clever change.

File: bitcash/transaction.py [hunk 3]
```diff
@@ -113,7 +113,10 @@ def sanitize_tx_data(unspents, outputs, fee, leftover, combine=True, message=Non
     messages = []
 
     if message:
-        message_chunks = chunk_data(message.encode('utf-8'), MESSAGE_LIMIT)
+        if not bytes:
```
Comment (teran-mckinney) [inline bitcash/transaction.py]: Isn't `bytes()` a bultin? Not sure if using bytes as a name will cause issues or just confusion.

File: bitcash/wallet.py [hunk 4]
```diff
@@ -278,8 +282,11 @@ def send(self, outputs, fee=None, leftover=None, combine=True,
         :type combine: ``bool``
         :param message: A message to include in the transaction. This will be
                         stored in the blockchain forever. Due to size limits,
-                        each message will be stored in chunks of 40 bytes.
+                        each message will be stored in chunks of 220 bytes.
         :type message: ``str``
+        :param bytes: Whether or not the supplied message is already bytes. If
+                      false (default), message is automatically encoded to UTF-8.
+        :type bytes: ``bool``
```
Comment (teran-mckinney) [inline bitcash/wallet.py]: I think maybe `type` could make more sense here, or would automatic type detection be the Pythonic way?

I guess maybe it's not a good one to make automatic, but I think a typed input seems ideal where it just takes both types and processes them appropriately.

File: bitcash/transaction.py (hunk 3 above)
Comment (lannocc) [reply bitcash/transaction.py]: Ugh, Python taking all the good names...

File: bitcash/wallet.py (hunk 4 above)
Comment (lannocc) [reply bitcash/wallet.py]: I did a little Google searching at the time and gave up and just added this flag instead. But I'm looking at it more right now. The Pythonic way (duck typing) would be not to check that it is a specific type but, in this case, probably to see that it has the encode() function, as so:

    try:
        message = message.encode('utf-8')
    except AttributeError:
        pass # assume message is already a bytes-like object

**Comment by lannocc** [comment]
I've committed your suggested changes @teran-mckinney. Should be a bit cleaner now.

**Comment by excalibur0922** [comment]
Cool thanks @teran-mckinney. When I get home from work (around 10 hrs from now) I'll tidy that right up (I'll just wipe my fork and start over with just those changes). 

I don't mind which you prefer to merge first :)

**Comment by teran-mckinney** [comment]
Thank you to both of you. This is looking great. Let me know if you want me to push  this out on pip soon or if I should wait a bit.

**Pull request merged**